src/string/extglobmatch.c

```diff
@@ -0,0 +1,463 @@
+                                                .repeat = 1,
+                                                .repeat_pat = eg.cur_pat,
+                                                .nested = con.nested + 1};
+                        res = glob_match(arg);
+                        if (res == 1) {
+                            return res;
+                        }
+                        eg = find_next(eg);
+                    }
+                } else {
+                    malformed_pattern(p);
+                }
+                return 0;
+            case '!':
+                /* Matches anything except one of the given patterns */
+                if (con.pat[0] == '(') {
+                    struct ext_glob eg = find_next((struct ext_glob)
+                                                        {.cur_pat = con.pat});
+                    if (c == '\0') /// TODO check this
+                        return 1 ^ con.except;
+                    int res;
+                    while (!eg.ret) {
+                        struct context arg = {  .str = con.str,
+                                                .pat = eg.cur_pat,
+                                                .pat_chunk_end = eg.end_pat,
+                                                .back_pat = back_pat,
+                                                .back_str = back_str,
+                                                .except = 1,
+                                                .nested = con.nested + 1,
+                                                .repeat = con.repeat,
+                                                .repeat_pat = con.repeat_pat};
+                        res = glob_match(arg);
+                        if (res == 1) {
+                            goto backtrack;
+                        }
+                        eg = find_next(eg);
+                    }
+                    con.pat = eg.end_pat;
+                    back_pat = con.pat;
+                    back_str = con.str;
+                } else {
+                    malformed_pattern(p);
+                }
+                break;
+            case '[': /* Character class */
+                {
+                    int match = 0, inverted = (*con.pat == '!');
+                    char const *class = con.pat + inverted;
+                    unsigned char a = *class++;
+                    /*
+                     * Iterate over each span in the character class.
+                     * A span is either a single character a, or a
+                     * range a-b.  The first span may begin with ']'.
+                     */
+                    do {
+                        unsigned char b = a;
+                        if (a == '\0')	/* Malformed */
+                            goto literal;
+                        if (class[0] == '-' && class[1] != ']') {
+                            b = class[1];
+                            if (b == '\0')
```

Comment:
I have some trouble understanding this particular check -- and even more so as the branch is never reached.

Since `b` is defined as `a`, and `a` is checked against `\0` @ line 358, this check is in fact redundant and should be removed.